File: mini-agent/internal/embed/embed.go
// Package embed 封装 OpenAI 兼容的 embedding API 客户端（默认指向硅基流动）。
//
// 在 agent 链路中的位置：这是 RAG 的"翻译器"——把文本翻译成向量，
// 下游的向量库（internal/vectorstore，练习 2）和 RAG 工具（练习 4）都依赖它。
//
// 为什么单独建包而不是复用 internal/llm：
// DeepSeek 官方没有 embedding API，embedding 必须走另一家服务商
// （硅基流动 / 本地 Ollama），baseURL、apiKey、模型名、端点（/embeddings）
// 全都不同，硬塞进 llm.Client 只会让它背两套配置。
// 两者相同的是"OpenAI 兼容协议"这个外壳：同样的 Authorization 头、
// 同样的错误处理模式（APIError）、同样的重试需求。
package embed

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mini-agent/internal/llm"
	"net/http"
	"strings"
	"time"
)

// bge-m3 的输出维度。写死它的意义：入库前校验向量长度，
// 维度错了说明模型/服务商配错了，越早报错越好——
// 否则错误向量悄悄入库，检索结果全错还很难排查。
const BgeM3Dimensions = 1024

// Client 是一个 OpenAI 兼容的 embeddings 客户端。
// 字段与 llm.Client 同构，学习时可以对照着看。
type Client struct {
	baseURL    string
	apiKey     string
	model      string
	httpClient *http.Client
}

// NewClient 默认指向硅基流动（有免费的 bge-m3，注册送额度）。
// apiKey 从环境变量 SILICONFLOW_API_KEY 读入后传入。
func NewClient(apiKey string) *Client {
	return &Client{
		baseURL: "https://api.siliconflow.cn/v1",
		apiKey:  apiKey,
		model:   "BAAI/bge-m3",
		httpClient: &http.Client{
			Timeout: 60 * time.Second, // embedding 是批处理接口，单批文本多时需要余量
		},
	}
}

// WithBaseURL 切换服务商，例如本地 Ollama："http://localhost:11434/v1"。
func (c *Client) WithBaseURL(baseURL string) *Client {
	c.baseURL = baseURL
	return c
}

// WithModel 切换 embedding 模型。
// 注意：换模型 = 换向量空间，已入库的旧向量全部作废，必须重建索引。
func (c *Client) WithModel(model string) *Client {
	c.model = model
	return c
}

// embeddingRequest 是发给 /embeddings 的请求体。
// input 是数组——批量接口，一次请求 embedding 多段文本，
// 比逐段调用省掉大量 HTTP 往返（入库几百个 chunk 时差距明显）。
type embeddingRequest struct {
	Model string   `json:"model"`
	Input []string `json:"input"`
}

// embeddingResponse 是 /embeddings 的响应体。
//
// 本包最核心的坑：data 数组的顺序不能假设与输入顺序一致！
// 每个元素带 index 字段标明它对应 input 的第几段，
// 归位时必须按 index 放，不能按数组下标直接对应。
// （部分服务商会按 token 数等内部策略重排，文档不保证顺序。）
type embeddingResponse struct {
	Data []struct {
		Index     int       `json:"index"`
		Embedding []float32 `json:"embedding"`
	} `json:"data"`
	Usage struct {
		PromptTokens int `json:"prompt_tokens"`
		TotalTokens  int `json:"total_tokens"`
	} `json:"usage"`
}

// Embed 输入一批文本，返回与之一一对应的向量切片（result[i] 是 texts[i] 的向量）。
// 实现要点见函数内注释；本练习的参考答案在
// docs/solutions/stage-02/exercise-1-embedding-client.md。
func (c *Client) Embed(texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, fmt.Errorf("embed: empty input")
	}

	for i, t := range texts {
		if strings.TrimSpace(t) == "" {
			return nil, fmt.Errorf("embed: texts[%d] is empty", i)
		}
	}

	body, err := json.Marshal(embeddingRequest{Model: c.model, Input: texts})
	if err != nil {
		return nil, fmt.Errorf("build request : %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, c.baseURL+"/embeddings", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("build request : %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.apiKey)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, &llm.APIError{StatusCode: resp.StatusCode, Body: string(respBody)}
	}

	var embResp embeddingResponse
	if err := json.Unmarshal(respBody, &embResp); err != nil {
		return nil, fmt.Errorf("unmarshal response :%w", err)
	}

	if len(embResp.Data) != len(texts) {
		return nil, fmt.Errorf("embed: got %d embeddings for %d texts", len(embResp.Data), len(texts))
	}

	result := make([][]float32, len(texts))

	for _, d := range embResp.Data {
		if d.Index < 0 || d.Index >= len(texts) {
			return nil, fmt.Errorf("embed: index %d out of range [0,%d]", d.Index, len(texts))
		}

		if len(d.Embedding) != BgeM3Dimensions {
			return nil, fmt.Errorf("embed: text[%d] dim = %d, want %d", d.Index, len(d.Embedding), BgeM3Dimensions)
		}
		result[d.Index] = d.Embedding
	}

	for i, v := range result {
		if v == nil {
			return nil, fmt.Errorf("embed: texts[%d] missing in response", i)
		}
	}

	return result, nil
}

File: stage-02-kb-agent/lib/embed.ts
/**
 * lib/embed.ts —— embedding 客户端：把文本变成向量（RAG 的"翻译器"）。
 *
 * 在 RAG 链路中的位置：chunking 之后、向量库之前；
 * 写入路径（embed 每个 chunk）和查询路径（embed 用户问题）共用本模块，
 * 两侧必须用同一个 embedding 模型，向量空间才对齐。
 *
 * 本模块与 Go 侧 mini-agent/internal/embed（练习 1）对应：
 * DeepSeek 官方没有 embedding API，embedding 必须走另一家服务商——
 * 这里直接 fetch 硅基流动的 OpenAI 兼容接口（POST /v1/embeddings，
 * 模型 BAAI/bge-m3），不引 SDK：批量接口就是一个 POST + JSON，
 * 手写几十行就能说清楚，引 SDK 反而遮住"按 index 归位"这个核心细节。
 */

/**
 * bge-m3 的输出维度。写死它的意义：入库前校验向量长度，
 * 维度错了说明模型/服务商配错了，越早报错越好——
 * 否则错误向量悄悄入库，检索结果全错还很难排查。
 */
export const BGE_M3_DIMENSIONS = 1024;

// TODO(练习6): 批量 embedding 调用 embedTexts
//
// 【任务】实现：
//
//   export async function embedTexts(texts: string[]): Promise<number[][]>
//
// 输入一批文本，返回与之一一对应的向量数组（result[i] 是 texts[i] 的向量）。
// 要求两条路径：
//
//  1. 真实路径：fetch https://api.siliconflow.cn/v1/embeddings
//     - 请求体 { model: "BAAI/bge-m3", input: texts }，
//       请求头 Authorization: `Bearer ${process.env.SILICONFLOW_API_KEY}`；
//     - 入参校验：texts 为空、任一段为空白、apiKey 缺失都要尽早报错；
//     - 响应非 200 时把状态码和响应体带进错误信息（排查限流/余额问题全靠它）；
//     - 核心坑（与 Go 侧相同）：响应 data 数组的顺序不能假设与输入一致！
//       每个元素带 index 字段标明对应 input 的第几段，归位必须按 index 放；
//       同时校验 index 不越界、每段向量维度 === BGE_M3_DIMENSIONS、
//       归位后无空洞（防止服务端漏返回）。
//
//  2. mock 路径：process.env.EMBEDDING_MOCK === "1" 时，不调用真实 API，
//     返回"确定性假向量"——同一段文本永远得到同一个向量
//     （例如用文本 hash 做种子驱动一个简单伪随机发生器，如 xorshift，
//     生成 BGE_M3_DIMENSIONS 维的向量）。
//     为什么要有这个开关（可测试性教学点）：embedding 是外部依赖，
//     没有 key / 没网 / 怕花钱时全链路就没法验证。用确定性的 fake 替换
//     外部依赖，ingest 流水线（chunk → embed → 入库 → 落盘）可以离线
//     端到端跑通；假向量没有语义，检索质量无意义，但"管线通不通"能验证。
//     这正是单测里 mock/stub 思想在手工脚本场景的应用。
//
// 【提示】
//   - Next.js Route Handler 运行在 Node.js 环境，全局 fetch 直接用，
//     不需要 node-fetch 之类的库。
//   - 批量一次请求全部 chunk，比逐段调用省掉大量 HTTP 往返；
//     真实项目里文档很大时要分批（硅基流动单批有上限），本练习可先不分批，
//     在注释里说明这个局限即可。
//
// 【验收】EMBEDDING_MOCK=1 时无需任何 key，curl 上传样例文档（见
// app/api/ingest/route.ts 的 TODO 验收）全链路跑通；有真实
// SILICONFLOW_API_KEY 时关掉 mock 再跑一次，返回的向量应为 1024 维。
//
// 参考答案：docs/solutions/stage-02/exercise-6-ingest-pipeline.md（完成后再看）
export async function embedTexts(texts: string[]): Promise<number[][]> {
  throw new Error("embedTexts: TODO(练习6) 未实现，见本函数上方注释");
}
